sql/sql_acl.cc

```diff
@@ -13059,36 +13018,51 @@ void Sql_cmd_grant::warn_hostname_requires_resolving(THD *thd,
 }
 
 
-void Sql_cmd_grant::grant_stage0(THD *thd)
+bool Sql_cmd_grant::grant_stage0(THD *thd)
 {
   thd->binlog_invoker(false);   // Replicate current user as grantor
   if (thd->security_ctx->user)  // If not replication
     warn_hostname_requires_resolving(thd, thd->lex->users_list);
-}
 
+  /*
+    Resolve all users (current_user, current_role,
+    user_without_host_specified).
 
-bool Sql_cmd_grant::user_list_reset_mqh(THD *thd, List<LEX_USER> &users)
-{
-  List_iterator <LEX_USER> it(users);
-  LEX_USER *user, *tmp_user;
-  while ((tmp_user= it++))
+    This resolved list is needed within all grant ops so we store it as
+    a class member.
+  */
+  List_iterator<LEX_USER> it(thd->lex->users_list);
+  m_resolved_users.empty();
+  for (LEX_USER *tmp_user= it++; tmp_user != NULL; tmp_user= it++)
   {
-    if (!(user= get_current_user(thd, tmp_user)))
+    LEX_USER *resolved_user= get_current_user(thd, tmp_user);
```

Comment:
ok, so you add an extra loop over users. Where do you eliminate a number of mallocs?